File: web/src/architecture/infrastructure/api/publicShare.ts
import { get, post } from '@/architecture/infrastructure/apiClient/request'
import type { FunctionDetail, FunctionSchema } from '@/architecture/domain/types'
import type { FormSubmitRequest, IFormGateway } from '@/architecture/domain/interfaces/IFormGateway'

const PUBLIC_ANONYMOUS_TOKEN_HEADER = 'X-Public-Anonymous-Token'
const PUBLIC_ANONYMOUS_TOKEN_STORAGE_KEY = 'public_anonymous_token'
let anonymousTokenPromise: Promise<string> | null = null

export interface PublicShareView {
  share_id: string
  title: string
  description?: string
  full_code_path: string
  schema: FunctionSchema
  expires_at?: string
  remaining_uses?: number
}

export interface PublicAnonymousToken {
  anonymous_token: string
  expires_at: string
}

export type PublicShareSubmitResult = unknown

export interface CreatePublicShareRequest {
  full_code_path: string
  title?: string
  description?: string
  expires_at?: string | null
  max_uses?: number
}

export interface PublicShareItem {
  share_id: string
  tenant_user: string
  app: string
  full_code_path: string
  resource_type: string
  action: string
  title: string
  description?: string
  enabled: boolean
  expires_at?: string
  max_uses: number
  use_count: number
  last_used_at?: string
  created_at: string
  created_by: string
  public_url?: string
}

export interface PublicShareList {
  items: PublicShareItem[]
}

export interface PublicShareListParams {
  full_code_path: string
  keyword?: string
  created_by?: string
  status?: string
}

function getAnonymousToken(): string {
  return localStorage.getItem(PUBLIC_ANONYMOUS_TOKEN_STORAGE_KEY) || ''
}

function saveAnonymousToken(token?: string) {
  if (token) {
    localStorage.setItem(PUBLIC_ANONYMOUS_TOKEN_STORAGE_KEY, token)
  }
}

export function publicShareAnonymousHeaders(): Record<string, string> {
  const token = getAnonymousToken()
  return token ? { [PUBLIC_ANONYMOUS_TOKEN_HEADER]: token } : {}
}

async function requestPublicAnonymousToken(): Promise<string> {
  const resp = await post<PublicAnonymousToken>('/public/api/anonymous-token', undefined, {
    headers: publicShareAnonymousHeaders(),
  })
  saveAnonymousToken(resp.anonymous_token)
  return resp.anonymous_token
}

export async function ensurePublicAnonymousToken(options: { refresh?: boolean } = {}): Promise<string> {
  const currentToken = getAnonymousToken()
  if (currentToken && !options.refresh) {
    return currentToken
  }

  if (!anonymousTokenPromise) {
    anonymousTokenPromise = requestPublicAnonymousToken().finally(() => {
      anonymousTokenPromise = null
    })
  }
  return anonymousTokenPromise
}

export function getCurrentPublicShareId(): string {
  const match = window.location.pathname.match(/^\/public\/s\/([^/]+)/)
  return match?.[1] ? decodeURIComponent(match[1]) : ''
}

export async function getPublicShareView(shareId: string): Promise<PublicShareView> {
  await ensurePublicAnonymousToken({ refresh: true })
  return get<PublicShareView>(`/public/api/s/${shareId}`, undefined, false, {
    headers: publicShareAnonymousHeaders(),
  })
}

export async function submitPublicShare(shareId: string, data: Record<string, unknown>): Promise<PublicShareSubmitResult> {
  await ensurePublicAnonymousToken()
  return post<PublicShareSubmitResult>(`/public/api/s/${shareId}/submit`, data, {
    headers: publicShareAnonymousHeaders(),
  })
}

export function createPublicShare(req: CreatePublicShareRequest): Promise<PublicShareItem> {
  return post<PublicShareItem>('/workspace/api/v1/public_shares', req)
}

export function listPublicShares(params: PublicShareListParams): Promise<PublicShareList> {
  return get<PublicShareList>('/workspace/api/v1/public_shares', params)
}

export function disablePublicShare(shareId: string): Promise<unknown> {
  return post(`/workspace/api/v1/public_shares/${shareId}/disable`)
}

export function createPublicShareFunctionDetail(view: PublicShareView): FunctionDetail {
  return {
    id: 0,
    method: 'POST',
    router: view.full_code_path,
    template_type: 'form',
    name: view.title,
    description: view.description,
    schema: view.schema,
  }
}

export class PublicShareFormGateway implements IFormGateway {
  constructor(private shareId: string) {}

  async submitForm(request: FormSubmitRequest): Promise<unknown> {
    const response = await submitPublicShare(this.shareId, request.data)
    return response && typeof response === 'object'
      ? response
      : { result: response }
  }
}

File: dto/public_share.go
package dto

import (
	"encoding/json"
	"time"

	"github.com/kageos/kageos/pkg/functionschema"
)

type CreatePublicShareReq struct {
	FullCodePath string          `json:"full_code_path" binding:"required"`
	Title        string          `json:"title,omitempty"`
	Description  string          `json:"description,omitempty"`
	ExpiresAt    *time.Time      `json:"expires_at,omitempty"`
	MaxUses      int             `json:"max_uses,omitempty"`
	PresetValues json.RawMessage `json:"preset_values,omitempty" swaggertype:"object"`
}

type PublicShareResp struct {
	ShareID      string          `json:"share_id"`
	TenantUser   string          `json:"tenant_user"`
	App          string          `json:"app"`
	FullCodePath string          `json:"full_code_path"`
	ResourceType string          `json:"resource_type"`
	Action       string          `json:"action"`
	Title        string          `json:"title"`
	Description  string          `json:"description"`
	Enabled      bool            `json:"enabled"`
	ExpiresAt    *time.Time      `json:"expires_at,omitempty"`
	MaxUses      int             `json:"max_uses"`
	UseCount     int             `json:"use_count"`
	LastUsedAt   *time.Time      `json:"last_used_at,omitempty"`
	CreatedAt    time.Time       `json:"created_at"`
	CreatedBy    string          `json:"created_by"`
	PublicURL    string          `json:"public_url,omitempty"`
	PresetValues json.RawMessage `json:"preset_values,omitempty" swaggertype:"object"`
}

type PublicShareListResp struct {
	Items []*PublicShareResp `json:"items"`
}

type PublicShareViewResp struct {
	ShareID       string                         `json:"share_id"`
	Title         string                         `json:"title"`
	Description   string                         `json:"description"`
	FullCodePath  string                         `json:"full_code_path"`
	Schema        *functionschema.FunctionSchema `json:"schema"`
	ExpiresAt     *time.Time                     `json:"expires_at,omitempty"`
	RemainingUses *int                           `json:"remaining_uses,omitempty"`
	PresetValues  json.RawMessage                `json:"preset_values,omitempty" swaggertype:"object"`
}

type PublicAnonymousTokenResp struct {
	AnonymousToken string    `json:"anonymous_token"`
	ExpiresAt      time.Time `json:"expires_at"`
}

// PublicShareSubmissionItem 是公开表单当前匿名访客可见的提交记录。
// 这里只返回展示所需的脱敏字段，不暴露租户、操作者、IP、User-Agent 等内部审计信息。
type PublicShareSubmissionItem struct {
	Status         string      `json:"status"`
	Summary        string      `json:"summary"`
	RequestBody    interface{} `json:"request_body,omitempty"`
	ResponseBody   interface{} `json:"response_body,omitempty"`
	DurationMillis int64       `json:"duration_millis,omitempty"`
	TraceID        string      `json:"trace_id,omitempty"`
	CreatedAt      string      `json:"created_at"`
}

type PublicShareSubmissionListResp struct {
	Items    []*PublicShareSubmissionItem `json:"items"`
	Total    int64                        `json:"total"`
	Page     int                          `json:"page"`
	PageSize int                          `json:"page_size"`
}
